### PR TITLE
Better volumes docs

### LINKED ISSUE #9915: Volume Documentation Needs to be Updated
[Volume documentation](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/volumes.md) only talks about these volumes:
* EmptyDir
* HostPath
* GCEPersistentDisk
* NFS

However, we [now support](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1/types.go#L211) many more volume types. Documentation should be updated to reflect this.

### PR DESCRIPTION
**Pull request by thockin**
Partially fixes #9915.  More to do, but I thought I'd put this up for review.  It could even get merged.  Will finish later or tomorrow.

### CODE REVIEW

**Comment by pmorie** [comment]
Really nice looking doc thockin. LGTM on the basic volume stuff but I don't know a lot about a PD claims yet. 

@markturansky TAL at persistent volume stuff?

**Comment by markturansky** [comment]
PVClaim part looks good to me.

**Comment by thockin** [comment]
pushed again, ready for review

**Comment by thockin** [comment]
@k8s-bot retest this please

**Comment by quinton-hoole** [comment]
FYI the e2e test flake is #10897, possibly related to #10868.

File: docs/volumes.md [hunk 1]
```diff
@@ -1,100 +1,182 @@
 # Volumes
-This document describes the current state of Volumes in kubernetes.  Familiarity with [pods](./pods.md) is suggested.
 
-A Volume is a directory, possibly with some data in it, which is accessible to a Container. Kubernetes Volumes are similar to but not the same as [Docker Volumes](https://docs.docker.com/userguide/dockervolumes/).
-
-A Pod specifies which Volumes its containers need in its [spec.volumes](http://kubernetes.io/third_party/swagger-ui/#!/v1/createPod) property.
-
-A process in a Container sees a filesystem view composed from two sources: a single Docker image and zero or more Volumes.  A [Docker image](https://docs.docker.com/userguide/dockerimages/) is at the root of the file hierarchy.  Any Volumes are mounted at points on the Docker image;  Volumes do not mount on other Volumes and do not have hard links to other Volumes.  Each container in the Pod independently specifies where on its image to mount each Volume.  This is specified in each container's VolumeMounts property.
+On-disk files in a container are ephemeral.  If that container crashes and
+restarts, the files are wiped, and the container starts with a clean slate.
```
Comment (brendandburns) [inline docs/volumes.md]: nit: s/wiped/lost/

Comment (brendandburns) [inline docs/volumes.md]: also clarify that the restart is external via kubelet.

File: docs/volumes.md [hunk 2]
```diff
@@ -1,100 +1,182 @@
...
+Running containers together in a pod, it is frequently necessary to share files
```
Comment (brendandburns) [inline docs/volumes.md]: Add a transition here:

"Additionally, running containers ..." ?

File: docs/volumes.md [hunk 3]
```diff
@@ -1,100 +1,182 @@
...
+between those containers.  The Kubernetes `volume` abstraction solves both of
```
Comment (brendandburns) [inline docs/volumes.md]: Capitalize 'Volume'

File: docs/volumes.md [hunk 4]
```diff
@@ -1,100 +1,182 @@
...
+Familiarity with [pods](./pods.md) is suggested.
+
+## Background
+
+Docker also has a concept of
+[volumes](https://docs.docker.com/userguide/dockervolumes/), though it is
+somewhat looser and less managed.  In Docker, a volume is simply a directory on
+disk or in another container.  Lifetimes are not managed and until very
+recently there were only local-disk-backed volumes.  Docker now provides volume
+drivers, but the functionality is very limited, still.
```
Comment (brendandburns) [inline docs/volumes.md]: Describe how?  Or cite issues (also, probably worth adding a date to this assertion, "As of 7/7/15 the functionality is very limited")

File: docs/volumes.md [hunk 5]
```diff
@@ -1,100 +1,182 @@
...
+
+A Kubernetes volume, on the other hand, has an explicit lifetime - the same as
+the pod that encloses it.  When a Pod ceases to exist, the volume will cease to
```
Comment (brendandburns) [inline docs/volumes.md]: "the pod that encloses it.  Consequently, it outlives an containers that run within the Pod, data is preserved across Container restarts.  Of course, when a Pod ceases to exist..."

File: docs/volumes.md [hunk 6]
```diff
@@ -1,100 +1,182 @@
...
+exist, too.  Perhaps more importantly than this, Kubernetes supports a number
```
Comment (brendandburns) [inline docs/volumes.md]: s/a number/many different/

File: docs/volumes.md [hunk 7]
```diff
@@ -1,100 +1,182 @@
...
+   * persistentVolumeClaim
+
+We welcome additional contributions.
+
+### emptyDir
+
+An `emptyDir` volume is first created when a Pod is assigned to a Node, and
+exists as long as that Pod is running on that node.  As the name says, it is
+initially empty.  Containers in the pod can all read and write the same
+files in the `emptyDir` volume.  When a Pod is removed from a node for any
```
Comment (brendandburns) [inline docs/volumes.md]: "volume, though it may be mounted in different locations in different containers."

File: docs/volumes.md [hunk 8]
```diff
@@ -1,100 +1,182 @@
...
+reason, the data in the `emptyDir` is deleted forever.  NOTE: a container
+crashing does *NOT* remove a pod from a node, so the data in an `emptyDir` volume
+is safe across container crashes.
+
+Some uses for an `emptyDir` are:
+
+* scratch space, such as for a disk-based mergesort or checkpointing a long
+  computation
```
Comment (brendandburns) [inline docs/volumes.md]: "checkpointing a long computation for recovery in case of crashes"

File: docs/volumes.md [hunk 9]
```diff
@@ -1,100 +1,182 @@
...
+### hostPath
+
+A `hostPath` volume mounts a file or directory from the host node's filesystem
+into your pod.  This is not something that most Pods will need, but it offers a
+powerful escape hatch for some applications.
+
+For example, some uses for a `hostPath` are:
+
+* running a container that needs access to Docker internals; use a `hostPath`
+  of /var/lib/docker
```
Comment (brendandburns) [inline docs/volumes.md]: code-font-ify /var/lib/docker

File: docs/volumes.md [hunk 10]
```diff
@@ -1,100 +1,182 @@
...
+* running cAdvisor in a container; use a `hostPath` of /dev/cgroups
```
Comment (brendandburns) [inline docs/volumes.md]: code-font-ify /dev/cgroups

**Comment by brendandburns** [comment]
Small nits, but LGTM.

Closes https://github.com/GoogleCloudPlatform/kubernetes/issues/9915

File: docs/volumes.md (hunk 1 above)
Comment (thockin) [inline docs/volumes.md]: done in next push

File: docs/volumes.md (hunk 2 above)
Comment (thockin) [inline docs/volumes.md]: retooled this pgph

File: docs/volumes.md (hunk 3 above)
Comment (thockin) [inline docs/volumes.md]: we're not always consistent about it, but sure - as an intro it makes sense.

File: docs/volumes.md (hunk 4 above)
Comment (thockin) [inline docs/volumes.md]: k

File: docs/volumes.md (hunk 5 above)
Comment (thockin) [inline docs/volumes.md]: k

File: docs/volumes.md (hunk 6 above)
Comment (thockin) [inline docs/volumes.md]: k

File: docs/volumes.md (hunk 7 above)
Comment (thockin) [inline docs/volumes.md]: k

File: docs/volumes.md (hunk 8 above)
Comment (thockin) [inline docs/volumes.md]: k

File: docs/volumes.md (hunk 9 above)
Comment (thockin) [inline docs/volumes.md]: k

File: docs/volumes.md (hunk 10 above)
Comment (thockin) [inline docs/volumes.md]: k

**Comment by thockin** [comment]
Comments fixed, thanks.

**Comment by thockin** [comment]
needs ok-to-merge

**Comment by thockin** [comment]
Adding my own OK :)  @vmarmol

**Pull request merged**